### PR TITLE
milvus supports partition operations

### PR DESCRIPTION
**Pull request by igerdy**
adds: 
 - The embedding store support inserting into specified partition.
  - The embedding store support searching from multiple partitions. 
 This feature is currently only supported by the milvus store. Milvus partitions belong to collections.

#518

### CODE REVIEW

File: langchain4j-core/src/main/java/dev/langchain4j/retriever/Retriever.java [hunk 1]
```diff
@@ -30,4 +30,18 @@ default List<T> findRelevant(
...
+     * Find relevant items for the given text and partitions.
+     * Currently only milvus is supported.
+     *
+     * <p>Default implementation throws an exception.
+     *
+     * @param text          the text to search for.
+     * @param partitions    milvus collection partition list.
+     * @return  the list of relevant items.
+     */
+    default List<T> findRelevant(String text, List<String> partitions) {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/retriever/Retriever.java]: I am not sure `Retriever` should be aware of partitions. `Retriever` is a much more generic concept that `EmbeddingStore`. `Retriever` can retrieve not only from `EmbeddingStore`, but from any data source. So the concept of partitions does not belong on this level.
`Retriever` is also supposed to be called in an automatic manned from `AiServices` or `ConversationalRetrievalChain`, both of which should be agnostic of partitions.

File: langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java [hunk 2]
```diff
@@ -113,4 +113,40 @@ default List<EmbeddingMatch<Embedded>> findRelevant(
...
+     * <p>Default implementation throws an exception.
+     *
+     * @param referenceEmbedding The embedding used as a reference. Returned embeddings should be relevant (closest) to this one.
+     * @param maxResults         The maximum number of embeddings to be returned.
+     * @param partitionNames     The range of partitions used for querying, Can be empty.
+     * @return A list of embedding matches.
+     * Each embedding match includes a relevance score (derivative of cosine distance),
+     * ranging from 0 (not relevant) to 1 (highly relevant).
+     */
+    default List<EmbeddingMatch<Embedded>> findRelevant(Embedding referenceEmbedding, int maxResults, List<String> partitionNames) {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java]: I would suggest placing this method on the `MilvusEmbeddingStore` as this is not necessarily applicable to all other implementations.

Comment (igerdy) [reply langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java:129]: I agree with your suggestion. 
My local version is under the milvus package: extending EmbeddingStore to MilvusEmbeddingStore, and then implementing MilvusEmbeddingStore.
Is this solution feasible? I'm doing local testing and will submit a PR after passing it.

File: langchain4j-core/src/main/java/dev/langchain4j/retriever/Retriever.java (hunk 1 above)
Comment (igerdy) [reply langchain4j-core/src/main/java/dev/langchain4j/retriever/Retriever.java:44]: I agree with your point of view.
I need to ask which of the following two options is more suitable? Are there any other more reasonable options?
In the Milvus package:
### options 1：

> step 1：

```
pom.xml add
        <dependency>
            <groupId>dev.langchain4j</groupId>
            <artifactId>langchain4j</artifactId>
        </dependency>
```
> step 2：
```
public interface MilvusRetrieverExtend<T> extends Retriever<T> {
    List<T> findRelevant(String text, List<String> partitions);
}
```
> step 3：
```
public class MilvusRetriever extends EmbeddingStoreRetriever implements MilvusRetrieverExtend<TextSegment> {
    ......
    public MilvusRetriever(EmbeddingStore<TextSegment> embeddingStore, EmbeddingModel embeddingModel, int maxResults, Double minScore) {
        super(embeddingStore, embeddingModel, maxResults, minScore);
    }

    @Override
    public List<TextSegment> findRelevant(String text, List<String> partitions) {
           ......
    }
    ......
}
```

### options 2：
> step 1：
```
public interface MilvusEmbeddingStoreRetrieverExtend<T> extends Retriever<T> {

    default List<T> findRelevant(String text) {
        return findRelevant(text, Collections.emptyList());
    }

    List<T> findRelevant(String text, List<String> partitions);

}
```
> step 2：
```
public class MilvusEmbeddingStoreRetriever implements MilvusEmbeddingStoreRetrieverExtend<TextSegment> {
    ......
    public MilvusEmbeddingStoreRetriever(EmbeddingStore<TextSegment> embeddingStore,
                                   EmbeddingModel embeddingModel,
                                   int maxResults,
                                   Double minScore) {
        this.embeddingStore = embeddingStore;
        this.embeddingModel = embeddingModel;
        this.maxResults = maxResults;
        this.minScore = minScore;
    }

    @Override
    public List<TextSegment> findRelevant(String text, List<String> partitions) {
           ......
    }
    ......
}

Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/retriever/Retriever.java]: I need to understand your use case better. How/where/when do you plan to decide which partition should be user for search? Will this be done statically or dynamically?

File: langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java (hunk 2 above)
Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java]: @igerdy yes, this makes sense!

Comment (igerdy) [reply langchain4j-core/src/main/java/dev/langchain4j/store/embedding/EmbeddingStore.java]: okay, I'll submit this part of the code first.

File: langchain4j-core/src/main/java/dev/langchain4j/retriever/Retriever.java (hunk 1 above)
Comment (igerdy) [reply langchain4j-core/src/main/java/dev/langchain4j/retriever/Retriever.java]: I only use milvus, and I think most people only use a certain store.
I hope milvus can be extended based on langchain, Standardize everyone’s usage habits.
For upper-layer applications, I hope to provide enough tool classes, but not pay too much attention to the specific implementation.
I plan to make it possible for each query to use a different partition range, or not specify partitions, by statically constructing and creating the collection join at query time.

**Pull request closed without merging**